File: backend/src/main/java/tom/projects/monkeylog/service/MeasurementService.java
package tom.projects.monkeylog.service;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import tom.projects.monkeylog.dto.measurement.MeasurementCreateRequest;
import tom.projects.monkeylog.dto.measurement.MeasurementPointCreateRequest;
import tom.projects.monkeylog.dto.measurement.MeasurementPointUpdateRequest;
import tom.projects.monkeylog.dto.measurement.MeasurementUpdateRequest;
import tom.projects.monkeylog.mapper.MeasurementMapper;
import tom.projects.monkeylog.model.measurement.Measurement;
import tom.projects.monkeylog.model.measurement.MeasurementPoint;
import tom.projects.monkeylog.repository.measurement.MeasurementPointRepository;
import tom.projects.monkeylog.repository.measurement.MeasurementRepository;
import tom.projects.monkeylog.security.AuthenticatedUser;

import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
@Service
public class MeasurementService {
    private static final String MEASUREMENT_NOT_FOUND = "Measurement not found.";
    private static final String MEASUREMENT_POINT_NOT_FOUND = "Measurement point not found.";
    private final MeasurementRepository measurementRepository;
    private final MeasurementPointRepository measurementPointRepository;
    private final MeasurementMapper measurementMapper;

    public List<Measurement> all() {
        return measurementRepository.findAll();
    }

    public Measurement get(UUID id) {
        return measurementRepository.findById(id)
                .filter(AuthenticatedUser::isResourceOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, MEASUREMENT_NOT_FOUND));
    }

    public MeasurementPoint getPoint(UUID id) {
        return measurementPointRepository.findById(id)
                .filter(AuthenticatedUser::isResourceOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, MEASUREMENT_POINT_NOT_FOUND));
    }

    public Measurement create(MeasurementCreateRequest measurementCreateRequest) {
        Measurement measurement = measurementMapper.measurementCreateRequestToMeasurement(measurementCreateRequest);
        measurement.setUserId(AuthenticatedUser.getId());

        return measurementRepository.save(measurement);
    }

    public Measurement update(MeasurementUpdateRequest measurementPatchDto, UUID id) {
        Measurement measurement = get(id);
        measurementMapper.updateMeasurement(measurement, measurementPatchDto);

        return measurementRepository.save(measurement);
    }

    public MeasurementPoint createPoint(UUID measurementId, MeasurementPointCreateRequest measurementPointCreateRequest) {
        Measurement measurement = get(measurementId);
        MeasurementPoint measurementPoint = new MeasurementPoint();
        measurementPoint.setValue(measurementPointCreateRequest.getValue());
        measurementPoint.setMeasurement(measurement);

        return measurementPointRepository.save(measurementPoint);
    }

    public MeasurementPoint updatePoint(UUID id, MeasurementPointUpdateRequest measurementPointUpdateRequest) {
        MeasurementPoint measurementPoint = getPoint(id);
        measurementPoint.setValue(measurementPointUpdateRequest.getValue());

        return measurementPointRepository.save(measurementPoint);
    }

    public void delete(UUID id) {
        measurementRepository.delete(get(id));
    }

    public void deletePoint(UUID id) {
        measurementPointRepository.delete(getPoint(id));
    }
}

File: backendkt/src/main/kotlin/tom/com/monkeylog/service/MeasurementService.kt
package tom.com.monkeylog.service

import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import tom.com.monkeylog.dto.measurement.MeasurementCreateRequest
import tom.com.monkeylog.dto.measurement.MeasurementPointCreateRequest
import tom.com.monkeylog.dto.measurement.MeasurementPointUpdateRequest
import tom.com.monkeylog.dto.measurement.MeasurementUpdateRequest
import tom.com.monkeylog.mapper.MeasurementMapper
import tom.com.monkeylog.model.measurement.Measurement
import tom.com.monkeylog.model.measurement.MeasurementPoint
import tom.com.monkeylog.repository.measurement.MeasurementPointRepository
import tom.com.monkeylog.repository.measurement.MeasurementRepository
import tom.com.monkeylog.security.AuthenticatedUser
import java.util.*

@Service
class MeasurementService(
    private val measurementRepository: MeasurementRepository,
    private val measurementPointRepository: MeasurementPointRepository,
    private val measurementMapper: MeasurementMapper,
) {
    fun all(): List<Measurement> {
        return measurementRepository.findAll()
    }

    operator fun get(id: UUID): Measurement {
        return measurementRepository.findById(id)
            .filter(AuthenticatedUser::isResourceOwner)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, MEASUREMENT_NOT_FOUND) }
    }

    fun getPoint(id: UUID): MeasurementPoint {
        return measurementPointRepository.findById(id)
            .filter(AuthenticatedUser::isResourceOwner)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, MEASUREMENT_POINT_NOT_FOUND) }
    }

    fun create(measurementCreateRequest: MeasurementCreateRequest): Measurement {
        val measurement: Measurement = measurementMapper.measurementCreateRequestToMeasurement(measurementCreateRequest)
        measurement.userId = AuthenticatedUser.id
        return measurementRepository.save(measurement)
    }

    fun update(measurementPatchDto: MeasurementUpdateRequest, id: UUID): Measurement {
        val measurement: Measurement = get(id)
        measurementMapper.updateMeasurement(measurement, measurementPatchDto)
        return measurementRepository.save(measurement)
    }

    fun createPoint(
        measurementId: UUID,
        measurementPointCreateRequest: MeasurementPointCreateRequest
    ): MeasurementPoint {
        val measurementPoint = MeasurementPoint(
            value = measurementPointCreateRequest.value,
            measurement = get(measurementId)
        )
        return measurementPointRepository.save(measurementPoint)
    }

    fun updatePoint(id: UUID, measurementPointUpdateRequest: MeasurementPointUpdateRequest): MeasurementPoint {
        val measurementPoint = getPoint(id)
        measurementPoint.value = measurementPointUpdateRequest.value
        return measurementPointRepository.save(measurementPoint)
    }

    fun delete(id: UUID) {
        measurementRepository.delete(get(id))
    }

    fun deletePoint(id: UUID) {
        measurementPointRepository.delete(getPoint(id))
    }

    companion object {
        private const val MEASUREMENT_NOT_FOUND = "Measurement not found."
        private const val MEASUREMENT_POINT_NOT_FOUND = "Measurement point not found."
    }
}
